Content.Server/NameIdentifier/NameIdentifierSystem.cs

```diff
@@ -0,0 +1,116 @@
+﻿using Content.Shared.GameTicking;
+using Content.Shared.NameIdentifier;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Random;
+
+namespace Content.Server.NameIdentifier;
+
+/// <summary>
+///     Handles unique name identifiers for entities e.g. `monkey (MK-912)`
+/// </summary>
+public class NameIdentifierSystem : EntitySystem
+{
+    [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+    [Dependency] private readonly IRobustRandom _robustRandom = default!;
+
+    [ViewVariables]
+    public Dictionary<NameIdentifierGroupPrototype, HashSet<int>> CurrentIds = new();
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<NameIdentifierComponent, ComponentInit>(OnComponentInit);
+        SubscribeLocalEvent<RoundRestartCleanupEvent>(CleanupIds);
+
+        InitialSetupPrototypes();
+        _prototypeManager.PrototypesReloaded += OnReloadPrototypes;
+    }
+
+    public override void Shutdown()
+    {
+        base.Shutdown();
+
+        _prototypeManager.PrototypesReloaded -= OnReloadPrototypes;
+    }
+
+    /// <summary>
+    ///     Generates a new unique name/suffix for a given entity and adds it to <see cref="CurrentIds"/>
+    ///     but does not set the entity's name.
+    /// </summary>
+    public string GenerateUniqueName(EntityUid uid, NameIdentifierGroupPrototype proto)
+    {
+        var entityName = Name(uid);
+        if (!CurrentIds.TryGetValue(proto, out var set))
+            return entityName;
+
+        if (set.Count == proto.MaxValue + 1)
+        {
+            // Oh jeez. We're outta numbers.
+            return entityName;
+        }
+
+        // This is kind of inefficient with very large amounts of entities but its better than any other method
+        // I could come up with.
+        var randomVal = _robustRandom.Next(0, proto.MaxValue);
```

Comment:
The correct solution here would be a non-repeating random sequence generator. An ""easy"" method would be a arbitrary period LFSR. This is someone else's problem though, I'm not sentencing you to figuring _that_ out.